Modules/fact_manager.py

```diff
@@ -0,0 +1,336 @@
+
+        See Fact class for more information on Fact properties (Modules\fact.py)
+
+        >>> example_query = sql.SQL("SELECT * some_table WHERE some_value=%s")
+        ... returned_fact = await self.query(example_query, ('something',))
+
+        Args:
+            name: name of fact to search, ie. 'prep'
+            lang: language ID for fact, defaults to 'en' (see Fact Class)
+
+        Returns: Fact()
+
+        """
+        # Build SQL Object for our query
+        query = sql.SQL(f"SELECT name, lang, message, aliases, author, edited, editedby, mfd from "
+                        f"{self._FACT_TABLE} where name=%s AND lang=%s")
+
+        # await our raw result from query
+        try:
+            rows = await self.query(query, (name, lang))
+        except (psycopg2.DatabaseError, psycopg2.ProgrammingError) as error:
+            # Check for offline database, or query errors
+            log.exception("Unable to find fact due to exception.")
+            raise
+
+        # unpack query into a fact object, or return an empty Fact if there is no result.
+        return Fact(*rows[0]) if rows else Fact()
+
+    async def log(self, fact_name: str, fact_lang: str, author: str, msg: str,
+                  new_field=None, old_field=None):
+        """
+        Writes a transaction log entry to the transaction log table.
+
+        The msg field should be only be one of the following (by convention):
+
+        * Added
+        * Deleted
+        * Edited
+        * Marked for delete
+        * Unmarked for delete
+
+        >>> await self.log("test", "en", "Shatt", "Marked for delete")
+
+        An Enum for these values is silly, as we need to output the strings directly
+        elsewhere, and so we do not preclude the use of future or custom strings from outside
+        mecha.
+
+        Args:
+            fact_name: Name of fact this applies to.
+            fact_lang: langID of fact this applies to.
+            author: the user that triggered the log entry.
+            msg: Message to be written to log.
+            new_field: Contents to place in the 'new' field for the log row.
+            old_field: Contents to place in the 'old' field for the log row.
+
+        Returns: Nothing.
+        """
+        log_query = sql.SQL(f"INSERT INTO {self._FACT_LOG} VALUES "
+                            f"(DEFAULT, %s, %s, %s, %s, %s, %s, %s)")
+
+        query_data = (fact_name, fact_lang, author, msg, old_field, new_field,
+                      datetime.datetime.utcnow())
+
+        try:
+            await self.query(log_query, query_data)
+        except (psycopg2.ProgrammingError, psycopg2.DatabaseError):
+            log.exception("Unable to write transaction log to table.")
+            raise
+
+    async def mfd(self, name: str, lang: str) -> bool:
+        """
+        Toggles the 'marked for deletion' flag on a fact.
+
+        >>> mfd_value = await self.mfd("test", "en")
```

Comment:
missing check on return value

Reply:
fixed in 83935a7

Reply:
appears not resolved, the doctest does not test `mfd_value`

Reply:
Fixed in 0d99f5d9